src/Interpreters/Cache/QueryResultCache.h

```diff
@@ -0,0 +1,160 @@
+/// which it becomes stale and is ignored. Stale entries are removed opportunistically from the cache, they are only evicted when a new
+/// entry is inserted and the cache has insufficient capacity.
+class QueryResultCache
+{
+public:
+    /// Represents a query result in the cache.
+    struct Key
+    {
+        /// ----------------------------------------------------
+        /// The actual key (data which gets hashed):
+
+        /// Unlike the query string, the AST is agnostic to lower/upper case (SELECT vs. select)
+        const ASTPtr ast;
+
+        /// Identifies a (virtual) cache partition. Can be used to cache the same query multiple times with different timeouts.
+        const String partition_key;
+
+        /// Note: For a transactionally consistent cache, we would need to include the system settings in the cache key or invalidate the
+        /// cache whenever the settings change. This is because certain settings (e.g. "additional_table_filters") can affect the query
+        /// result.
+
+        /// ----------------------------------------------------
+        /// Additional stuff data stored in the key, not hashed:
+
+        /// For constructing the pipe.
+        const Block header;
+
+        /// Std::nullopt means that the associated entry can be read by other users. In general, sharing is a bad idea: First, it is
+        /// unlikely that different users pose the same queries. Second, sharing potentially breaches security. E.g. User A should not be
+        /// able to bypass row policies on some table by running the same queries as user B for whom no row policies exist.
+        const std::optional<String> username;
+
+        /// When does the entry expire?
+        const std::chrono::time_point<std::chrono::system_clock> expires_at;
+
+        Key(ASTPtr ast_, String partition_key_,
+            Block header_, const std::optional<String> & username_,
+            std::chrono::time_point<std::chrono::system_clock> expires_at_);
+
+        bool operator==(const Key & other) const;
+        String queryStringFromAst() const;
+    };
+
+private:
+    struct KeyHasher
+    {
+        size_t operator()(const Key & key) const;
+    };
+
+    /// query --> query result
+    using Cache = std::unordered_map<Key, Chunk, KeyHasher>;
+
+    /// query --> query execution count
+    using TimesExecuted = std::unordered_map<Key, size_t, KeyHasher>;
+
+public:
+    /// Buffers multiple partial query result chunks (buffer()) and eventually stores them as cache entry (finalizeWrite()).
+    ///
+    /// Implementation note: Queries may throw exceptions during runtime, e.g. out-of-memory errors. In this case, no query result must be
+    /// written into the query result cache. Unfortunately, neither the Writer nor the special transform added on top of the query pipeline
+    /// which holds the Writer know whether they are destroyed because the query ended successfully or because of an exception (otherwise,
```

Comment:
what if query will be cancelled (e.g. by ctrl+c in terminal)?

Reply:
Ah, thanks, I missed that but it's fixed now.